File: packages/core-ts/src/registry.ts
/**
 * Session-scoped source registry.
 *
 * A `sourceId` is an opaque capability, not an address: minted per session, carrying the
 * authorization decision made when the source was registered, and never resolving in
 * another session. Expired handles are refused rather than silently re-fetched.
 */
import { randomBytes } from "node:crypto";

import { ShuntError } from "./errors.js";
import { DEFAULT_LIMITS, Limits } from "./limits.js";
import { Snapshot } from "./snapshot.js";

export interface RegisteredSource {
  readonly sourceId: string;
  readonly sessionId: string;
  readonly snapshot: Snapshot;
  readonly expiresAtEpoch: number;
  readonly internal: boolean;
}

function mintId(): string {
  return "src_" + randomBytes(8).toString("hex");
}

export class SourceRegistry {
  private readonly bySession = new Map<string, Map<string, RegisteredSource>>();

  constructor(
    private readonly limits: Limits = DEFAULT_LIMITS,
    private timeFn: () => number = () => Date.now() / 1000,
  ) {}

  /** Test seam for TTL assertions; production always uses the wall clock. */
  setTimeFn(fn: () => number): void {
    this.timeFn = fn;
  }

  register(sessionId: string, snapshot: Snapshot, internal = false): RegisteredSource {
    if (!sessionId) throw new ShuntError("UNSAFE_SOURCE", "NO_SESSION");
    const entry: RegisteredSource = {
      sourceId: mintId(),
      sessionId,
      snapshot,
      expiresAtEpoch: this.timeFn() + this.limits.spillTtlSeconds,
      internal,
    };
    let bucket = this.bySession.get(sessionId);
    if (!bucket) {
      bucket = new Map();
      this.bySession.set(sessionId, bucket);
    }
    bucket.set(entry.sourceId, entry);
    return entry;
  }

  resolve(sessionId: string, sourceId: string): RegisteredSource {
    const entry = this.bySession.get(sessionId)?.get(sourceId);
    if (!entry) {
      // A handle from another session is indistinguishable from an unknown one, which is
      // deliberate: cross-session probing learns nothing.
      throw new ShuntError("SOURCE_EXPIRED", "UNKNOWN_HANDLE");
    }
    if (this.timeFn() >= entry.expiresAtEpoch) {
      throw new ShuntError("SOURCE_EXPIRED", "TTL_ELAPSED");
    }
    return entry;
  }

  /** Registry-verified recursion guard. A payload claiming `internal` proves nothing. */
  isInternal(sessionId: string, sourceId: string): boolean {
    try {
      return this.resolve(sessionId, sourceId).internal;
    } catch {
      return false;
    }
  }

  expireSession(sessionId: string): number {
    const removed = this.bySession.get(sessionId)?.size ?? 0;
    this.bySession.delete(sessionId);
    return removed;
  }

  count(sessionId?: string): number {
    if (sessionId === undefined) {
      let total = 0;
      for (const bucket of this.bySession.values()) total += bucket.size;
      return total;
    }
    return this.bySession.get(sessionId)?.size ?? 0;
  }
}

File: packages/core-py/src/context_shunt/registry.py
"""Session-scoped source registry.

A ``source_id`` is an opaque capability, not an address: it is minted per session, it
carries the authorization decision that was made when the source was registered, and it
never resolves in another session. Expired handles are refused rather than silently
re-fetching the underlying file.
"""

from __future__ import annotations

import secrets
import threading
import time
from dataclasses import dataclass

from .errors import ShuntError
from .limits import DEFAULT_LIMITS, Limits
from .snapshot import Snapshot


@dataclass(frozen=True)
class RegisteredSource:
    source_id: str
    session_id: str
    snapshot: Snapshot
    expires_at_epoch: float
    internal: bool = False


def _mint_id() -> str:
    return "src_" + secrets.token_hex(8)


class SourceRegistry:
    """Thread-safe registry isolated per session/tenant."""

    def __init__(self, limits: Limits = DEFAULT_LIMITS, time_fn=time.time):
        self._limits = limits
        self._time = time_fn
        self._lock = threading.RLock()
        self._by_session: dict[str, dict[str, RegisteredSource]] = {}

    def register(
        self, session_id: str, snapshot: Snapshot, *, internal: bool = False
    ) -> RegisteredSource:
        if not session_id:
            raise ShuntError("UNSAFE_SOURCE", "NO_SESSION")
        entry = RegisteredSource(
            source_id=_mint_id(),
            session_id=session_id,
            snapshot=snapshot,
            expires_at_epoch=self._time() + self._limits.spill_ttl_seconds,
            internal=internal,
        )
        with self._lock:
            self._by_session.setdefault(session_id, {})[entry.source_id] = entry
        return entry

    def resolve(self, session_id: str, source_id: str) -> RegisteredSource:
        with self._lock:
            entry = self._by_session.get(session_id, {}).get(source_id)
        if entry is None:
            # A handle from another session is indistinguishable from an unknown one,
            # which is deliberate: cross-session probing learns nothing.
            raise ShuntError("SOURCE_EXPIRED", "UNKNOWN_HANDLE")
        if self._time() >= entry.expires_at_epoch:
            raise ShuntError("SOURCE_EXPIRED", "TTL_ELAPSED")
        return entry

    def is_internal(self, session_id: str, source_id: str) -> bool:
        """Registry-verified recursion guard. A payload claiming ``internal`` proves nothing."""
        try:
            return self.resolve(session_id, source_id).internal
        except ShuntError:
            return False

    def expire_session(self, session_id: str) -> int:
        with self._lock:
            removed = self._by_session.pop(session_id, {})
        return len(removed)

    def sweep(self) -> int:
        now = self._time()
        removed = 0
        with self._lock:
            for session_id, entries in list(self._by_session.items()):
                for source_id, entry in list(entries.items()):
                    if now >= entry.expires_at_epoch:
                        del entries[source_id]
                        removed += 1
                if not entries:
                    del self._by_session[session_id]
        return removed

    def count(self, session_id: str | None = None) -> int:
        with self._lock:
            if session_id is None:
                return sum(len(v) for v in self._by_session.values())
            return len(self._by_session.get(session_id, {}))
